rasa/core/channels/audiocodes.py

```diff
@@ -0,0 +1,250 @@
+        logger.debug(
+            f"Performing clean old conversations, current number: {len(self.conversations)}"
+        )
+        now = datetime.datetime.utcnow()
+        delta = datetime.timedelta(seconds=KEEP_ALIVE_SECONDS * 1.5)
+        self.conversations = {
+            k: v for k, v in self.conversations.items() if v.is_valid(now, delta)
+        }
+
+    def get_metadata(self, request: Request) -> Optional[Dict[Text, Text]]:
+        body = request.json
+        return body.get("activities", [{}])[0].get("parameters")
+
+    def blueprint(
+        self, on_new_message: Callable[[UserMessage], Awaitable[Any]]
+    ) -> Blueprint:
+        ac_webhook = Blueprint("ac_webhook", __name__)
+
+        @ac_webhook.route("/", methods=["GET"])
+        async def health(request: Request) -> HTTPResponse:
+            return response.json({"status": "ok"})
+
+        # {"conversation": <cid>, id, timestamp}
+        @ac_webhook.route("/webhook", methods=["POST"])
+        async def receive(request: Request) -> HTTPResponse:
+            if not self.scheduler_job:
+                await self._set_scheduler_job()
+            self._check_token(request.headers.get("Authorization"))
+            body = request.json
+            cid = body["conversation"]
+            if cid in self.conversations:
+                sanic.exceptions.abort(500, "Conversation already exists")
+            self.conversations[cid] = AudiocodesInput.Conversation(cid)
+            res = {
+                "activitiesURL": f"/webhooks/audiocodes/conversation/{cid}/activities",
+                "disconnectURL": f"/webhooks/audiocodes/conversation/{cid}/disconnect",
+                "refreshURL": f"/webhooks/audiocodes/conversation/{cid}/keepalive",
+                "expiresSeconds": KEEP_ALIVE_SECONDS,
+            }
+            return response.json(res)
+
+        # {"conversation": <cid>, "activities": List[Activity]}
+        @ac_webhook.route("/conversation/<cid>/activities", methods=["POST"])
+        async def on_activities(request: Request, cid: Text) -> HTTPResponse:
+            conversation = self._get_conversation(
+                request.headers.get("Authorization"), cid
+            )
+            metadata = self.get_metadata(request)
+            ac_output = AudiocodesOutput()
+            await conversation.handleActivities(
+                request.json,
+                output_channel=ac_output,
+                metadata=metadata,
+                on_new_message=on_new_message,
+            )
+            return response.json(
+                {"conversation": cid, "activities": ac_output.messages}
+            )
+
+        # {"conversation": <cid>, "reason": Optional[Text]}
+        @ac_webhook.route("/conversation/<cid>/disconnect", methods=["POST"])
+        async def disconnect(request: Request, cid: Text) -> HTTPResponse:
+            self._get_conversation(request.headers.get("Authorization"), cid)
+            # should add intent /end?
```

Comment:
We'd like to get your input on that. Thanks!

Reply:
Rasa Open Source doesn't mind if the user doesn't come back. So I'd say it's up to you.